### PR TITLE
Enable pyramid in advection example

### LINKED ISSUE #390: Advection example with pyramids
I have tried to run the advection example with a hybrid mesh consisting of hexes, tets and pyramids. This results in a whole cascade of asserts to trigger. Maybe someone with more knowledge about ghosts and pyramids can have a look at this. I have created a minimal testing example:
[cube.zip](https://github.com/DLR-AMR/t8code/files/10267294/cube.zip)
I have tried to run it with:
`mpirun -np 8 ./t8_advection -u7 -l0 -r1 -f cube -d3 -C0.5 -b4 -X-0.7 -Y0 -Z0 -R0.3 -T1` in debug mode.

### PR DESCRIPTION
**Pull request by holke**
**_Describe your changes here:_**

The advection example did not run with pyramids.
This PR should fix it.

Add pyramid cube option.
Fix two assertions that were too stiff
to allow for the flexible child count of pyramids. Fix face neighbor count that can now change from one level to the next.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

- [x] The author added a BSD statement to `doc/` (or already has one)
- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [x] All tests pass (in various configurations, this should be executed automatically in a github action)
- [x] New source/header files are properly added to the Makefiles
- [x] New Datatypes are added to t8indent_custom_datatypes.txt
- [x] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [x] New tests use the Google Test framework
- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [x] Testing of this template: If you feel something is missing from this list, contact the developers

### CODE REVIEW

**Comment by holke** [comment]
Calling `./t8_advection -u7 -l0 -r1 -f cube -d3 -C0.5 -b4 -X-0.7 -Y0 -Z0 -R0.3 -T1` results in an error since one tree with negative volume is detected.
The strange thing is, that this happens after the cmesh is read in, all negative volumes are corrected and the cmesh is repartitioned.
Usually, the volume should not change anymore after the reading is done.

File: src/t8_forest/t8_forest_cxx.cxx [hunk 1]
```diff
@@ -1929,6 +1929,7 @@ t8_forest_leaf_face_neighbors (t8_forest_t forest, t8_locidx_t ltreeid,
         element_index =
           t8_forest_bin_search_lower (element_array, neigh_id,
                                       forest->maxlevel);
+        T8_ASSERT (element_index >= 0);
```
Comment (sandro-elsweijer) [inline src/t8_forest/t8_forest_cxx.cxx:1932]: The fixed volume bug seems to have had an influence on this issue, but it is still not running. Now this assertion is triggering

Comment (holke) [reply src/t8_forest/t8_forest_cxx.cxx:1932]: Thanks.
In which time step did this occur and where you computing in parallel?

I am currently having a serial run that is in time step 6 and no assertion has occured so far.

Comment (holke) [reply src/t8_forest/t8_forest_cxx.cxx:1932]: I have merged with main which did a lot of updates, maybe this fixed the error that you got?

Comment (holke) [reply src/t8_forest/t8_forest_cxx.cxx:1932]: Update: Now in time step 20 and still going.
(due to debugging mode being on it is rather slow)

Comment (sandro-elsweijer) [reply src/t8_forest/t8_forest_cxx.cxx:1932]: I do not know the time step, but the assertion triggered after less than 10 sec of runtime in debug mode (I called `mpirun -np 8 --oversubscribe ./t8_advection -u7 -l0 -r1 -f cube -d3 -C0.5 -b4 -X-0.7 -Y0 -Z0 -R0.3 -T1`, `--oversubscribe` was needed due to me using OpenMPI)

Comment (sandro-elsweijer) [reply src/t8_forest/t8_forest_cxx.cxx:1932]: But with `mpirun -np 4 ./t8_advection -u7 -l0 -r1 -f cube -d3 -C0.5 -b4 -X-0.7 -Y0 -Z0 -R0.3 -T1` it runs smoothly

Comment (sandro-elsweijer) [reply src/t8_forest/t8_forest_cxx.cxx:1932]: Traceback:
```
[libsc 6] Abort: Assertion 'element_index >= 0'
[libsc 6] Abort: ./../t8code/src/t8_forest/t8_forest_cxx.cxx:1932
[libsc 6] Abort: Obtained 10 stack frames
[libsc 6] Stack 0: t8_advection(+0xc97a5) [0x558f0443b7a5]
[libsc 6] Stack 1: t8_advection(+0xc9760) [0x558f0443b760]
[libsc 6] Stack 2: t8_advection(+0xc9981) [0x558f0443b981]
[libsc 6] Stack 3: t8_advection(+0x5c539) [0x558f043ce539]
[libsc 6] Stack 4: t8_advection(+0x18a3c) [0x558f0438aa3c]
[libsc 6] Stack 5: t8_advection(+0x1963e) [0x558f0438b63e]
[libsc 6] Stack 6: t8_advection(+0x1b256) [0x558f0438d256]
[libsc 6] Stack 7: libc.so.6(+0x29d90) [0x7f6bd6891d90]
[libsc 6] Stack 8: libc.so.6(__libc_start_main+0x80) [0x7f6bd6891e40]
[libsc 6] Stack 9: t8_advection(+0x15d25) [0x558f04387d25]
```
Lines:
```
addr2line -e ./t8_advection -i
0x1b256
/home/elsw_sa/git/t8code_debug/./../t8code/example/advect/t8_advection.cxx:1851
0x1963e
/home/elsw_sa/git/t8code_debug/./../t8code/example/advect/t8_advection.cxx:1371
0x18a3c
/home/elsw_sa/git/t8code_debug/./../t8code/example/advect/t8_advection.cxx:1166
0x5c539
/home/elsw_sa/git/t8code_debug/./../t8code/src/t8_forest/t8_forest_cxx.cxx:1935
0xc9981
/home/elsw_sa/git/t8code_debug/sc/.././../t8code/sc/src/sc.c:1044
0xc9760
/home/elsw_sa/git/t8code_debug/sc/.././../t8code/sc/src/sc.c:982
0xc97a5
/home/elsw_sa/git/t8code_debug/sc/.././../t8code/sc/src/sc.c:997
```

Comment (holke) [reply src/t8_forest/t8_forest_cxx.cxx:1932]: This is not solved. But close the conversation for this PR.
See #390, which remains open.

**Comment by sandro-elsweijer** [comment]
We should check, if the bug is completely fixed in combination with this PR: https://github.com/DLR-AMR/t8code/pull/770

**Comment by ackirby88** [comment]
> We should check, if the bug is completely fixed in combination with this PR: #770  

Unfortunately, issue #774 demonstrates the #770 still does not resolve negative volumes for adaption. But I think the example in 774 will help us figure out the cause of negative volumes for refining tets, prisms, and/or pyramids.

**Review by lukasdreyer — APPROVED** [review]

**Pull request merged**